File: src/main/java/eladkay/quaritum/common/crafting/recipes/RecipeAwakenedSoulstone.java
package eladkay.quaritum.common.crafting.recipes;

import com.google.common.collect.Lists;
import eladkay.quaritum.api.animus.IFlower;
import eladkay.quaritum.common.item.soulstones.ItemAwakenedSoulstone;
import eladkay.quaritum.common.item.soulstones.ItemDormantSoulstone;
import net.minecraft.block.Block;
import net.minecraft.inventory.InventoryCrafting;
import net.minecraft.item.ItemStack;
import net.minecraft.item.crafting.IRecipe;
import net.minecraft.world.World;
import net.minecraftforge.common.ForgeHooks;

import javax.annotation.Nonnull;
import java.util.List;

public class RecipeAwakenedSoulstone implements IRecipe {
    @Override
    public boolean matches(@Nonnull InventoryCrafting inventoryCrafting, @Nonnull World world) {
        boolean foundSoulstone = false;
        for (int index = 0; index < inventoryCrafting.getSizeInventory(); ++index) {
            ItemStack stack = inventoryCrafting.getStackInSlot(index);
            if (stack != null && stack.getItem() instanceof ItemDormantSoulstone) foundSoulstone = true;
        }
        return foundSoulstone;
    }

    @Override
    public ItemStack getCraftingResult(@Nonnull InventoryCrafting inventoryCrafting) {
        int animus = 0;
        int rarity = Integer.MAX_VALUE;
        for (int index = 0; index < inventoryCrafting.getSizeInventory(); ++index) {
            ItemStack stack = inventoryCrafting.getStackInSlot(index);
            if (stack != null && Block.getBlockFromItem(stack.getItem()) instanceof IFlower) {
                animus += ((IFlower) Block.getBlockFromItem(stack.getItem())).getAnimusFromStack(stack);
                rarity = Math.min(((IFlower) Block.getBlockFromItem(stack.getItem())).getRarity(stack), rarity);
            }

        }
        return ItemAwakenedSoulstone.withAnimus(animus, rarity);
    }

    @Override
    public int getRecipeSize() {
        return 10;
    }

    @Override
    public ItemStack getRecipeOutput() {
        return null;
    }

    @Nonnull
    @Override
    public ItemStack[] getRemainingItems(@Nonnull InventoryCrafting inventoryCrafting) {
        return ForgeHooks.defaultRecipeGetRemainingItems(inventoryCrafting);
    }
}

File: src/main/java/eladkay/quaeritum/common/crafting/recipes/RecipeAwakenedSoulstone.kt
package eladkay.quaeritum.common.crafting.recipes

import com.teamwizardry.librarianlib.features.kotlin.isNotEmpty
import eladkay.quaeritum.api.animus.IAnimusResource
import eladkay.quaeritum.common.item.soulstones.ItemAwakenedSoulstone
import eladkay.quaeritum.common.item.soulstones.ItemDormantSoulstone
import net.minecraft.inventory.InventoryCrafting
import net.minecraft.item.ItemStack
import net.minecraft.item.crafting.IRecipe
import net.minecraft.util.NonNullList
import net.minecraft.world.World
import net.minecraftforge.common.ForgeHooks
import net.minecraftforge.registries.IForgeRegistryEntry

class RecipeAwakenedSoulstone : IForgeRegistryEntry.Impl<IRecipe>(), IRecipe {
    override fun matches(inventoryCrafting: InventoryCrafting, world: World): Boolean {
        var foundSoulstone = false
        var foundResource = false
        (0 until inventoryCrafting.sizeInventory)
                .map { inventoryCrafting.getStackInSlot(it) }
                .filter { it.isNotEmpty }
                .forEach {
                    if (it.item is ItemDormantSoulstone && !foundSoulstone) foundSoulstone = true
                    else if (it.item is IAnimusResource && (it.item as IAnimusResource).getAnimus(it) > 0) foundResource = true
                    else return false
                }

        return foundSoulstone && foundResource
    }

    override fun getCraftingResult(inventoryCrafting: InventoryCrafting): ItemStack? {
        var animus = 0
        var rarity = Integer.MAX_VALUE
        for (index in 0 until inventoryCrafting.sizeInventory) {
            val stack = inventoryCrafting.getStackInSlot(index)
            if (stack != null && stack.item is IAnimusResource) {
                animus += (stack.item as IAnimusResource).getAnimus(stack)
                rarity = Math.min((stack.item as IAnimusResource).getAnimusTier(stack).ordinal, rarity)
            }
        }
        return ItemAwakenedSoulstone.withAnimus(animus, rarity)
    }

    override fun canFit(width: Int, height: Int): Boolean {
        return true
    }

    override fun getRecipeOutput(): ItemStack {
        return ItemStack.EMPTY
    }

    override fun isDynamic(): Boolean {
        return true
    }

    override fun getRemainingItems(inventoryCrafting: InventoryCrafting): NonNullList<ItemStack> {
        return ForgeHooks.defaultRecipeGetRemainingItems(inventoryCrafting)
    }
}
